### PR TITLE
Part 2: Decimal to Binary conversion

### PR DESCRIPTION
**Pull request by wildmaples**
# Changes
* Parse booleans, new arithmetic calls and argument variables
* Translate while loops and if-else statement to VM code

### CODE REVIEW

File: lib/expression_parser.rb [hunk 1]
```diff
@@ -18,6 +18,13 @@ def write_vm_code(vm_writer)
   end
 end
 
+Boolean = Struct.new(:value) do
+  def write_vm_code(vm_writer)
+    vm_writer.write_push(:CONST, 0)
+    vm_writer.write_arithmetic(:NOT) if value == :TRUE
```
Comment (tomstuart) [inline lib/expression_parser.rb:32]: Yeah, I guess this is neater than having to conditionally `push constant 1` and then `neg`.

**Review by tomstuart — APPROVED** [review]
Excellent! This all seems correct to me, notwithstanding my nitpicks about naming. The end is in sight. 😅 

I do think it’d be enlightening for you to refactor `ExpressionParser` to no longer rely upon the symbol table, since (per my comment) it doesn’t really need to and the whole setup is simpler without it, but I appreciate that refactoring won’t be your top priority at the moment so I won’t complain if we skip it for now.

File: lib/expression_parser.rb (hunk 1 above)
Comment (wildmaples) [reply lib/expression_parser.rb:32]: Yeah my first instinct wasn't the best solution. Using `push constant 0` then `not` makes it extra clear that it's a `true`.

File: lib/expression_parser.rb [hunk 2]
```diff
@@ -52,9 +52,10 @@ def write_vm_code(vm_writer)
   end
 end
 
-LocalVariable = Struct.new(:name, :index) do
+IDENTIFIER = Struct.new(:name, :kind, :index) do
```
[the remark this answers is not in the document]
Comment (wildmaples) [reply lib/expression_parser.rb:57]: Ugh sorry, my brain was fried when I wrote this -- didn't mean to all caps it. FWIW, I went with `Variable` first, and decided not to for some reason.

**Pull request merged**